File: 0075_sort_colors/sort_colors.go
package main

/**
 * Time:    O(n)
 * Space:   O(1)
 */
func sortColors(nums []int) {
	c0, c1, c2 := 0, 0, len(nums)-1
	for c1 <= c2 {
		if nums[c1] == 0 {
			nums[c0], nums[c1] = nums[c1], nums[c0]
			c0++
			c1++
		} else if nums[c1] == 2 {
			nums[c2], nums[c1] = nums[c1], nums[c2]
			c2--
		} else {
			c1++
		}
	}
}

func main() {

}

File: 0075_sort_colors/sort_colors.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public void sortColors(int[] nums) {
        int c = 0, c0 = 0;
        int c2 = nums.length - 1;
        while (c <= c2) {
            if (nums[c] == 0) {
                swap(nums, c, c0);
                c++;
                c0++;
            } else if (nums[c] == 1) {
                c++;
            } else {
                swap(nums, c, c2);
                c2--;
            }
        }
    }
    
    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}